File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/common/uicomponent/music/service/MusicService.java
package com.trtc.uikit.livekit.common.uicomponent.music.service;

import static com.trtc.uikit.livekit.common.uicomponent.music.store.MusicPanelState.MusicInfo.INVALID_ID;

import com.tencent.liteav.audio.TXAudioEffectManager;
import com.tencent.trtc.TRTCCloud;
import com.trtc.uikit.livekit.common.uicomponent.music.store.MusicPanelSateFactory;
import com.trtc.uikit.livekit.common.uicomponent.music.store.MusicPanelState;
import com.trtc.uikit.livekit.common.utils.LiveKitLog;

public class MusicService {
    public static final String TAG = "MusicService";

    private final String          mRoomId;
    private final TRTCCloud       mTRTCCloud;
    public        MusicPanelState mMusicPanelState;

    public MusicService(String roomId, TRTCCloud trtcCloud) {
        mRoomId = roomId;
        mTRTCCloud = trtcCloud;
        mMusicPanelState = MusicPanelSateFactory.getState(roomId);
    }

    public void operatePlayMusic(MusicPanelState.MusicInfo musicInfo) {
        MusicPanelState.MusicInfo currentMusicInfo = mMusicPanelState.currentMusicInfo.get();
        if (currentMusicInfo.id != INVALID_ID && currentMusicInfo.id != musicInfo.id) {
            stopMusic(currentMusicInfo.id);
        }
        mMusicPanelState.currentMusicInfo.set(musicInfo);
        boolean isPlaying = musicInfo.isPlaying.get();
        LiveKitLog.info(TAG + "operatePlayMusic:[isPlaying:" + isPlaying + "]");
        if (musicInfo.isPlaying.get()) {
            stopMusic(musicInfo.id);
        } else {
            startMusic(musicInfo);
        }
    }

    private void startMusic(MusicPanelState.MusicInfo musicInfo) {
        LiveKitLog.info(TAG + "[" + mRoomId + "] startMusic:[musicInfo:" + musicInfo + "]");
        int id = musicInfo.id;
        String path = musicInfo.path;
        TXAudioEffectManager.AudioMusicParam audioMusicParam = new TXAudioEffectManager.AudioMusicParam(id, path);
        audioMusicParam.loopCount = Integer.MAX_VALUE;
        audioMusicParam.publish = true;
        mTRTCCloud.getAudioEffectManager().startPlayMusic(audioMusicParam);
        mTRTCCloud.getAudioEffectManager().setMusicPitch(id, musicInfo.pitch.get());
        mMusicPanelState.currentMusicInfo.get().isPlaying.set(true);
    }

    private void stopMusic(int id) {
        LiveKitLog.info(TAG + "[" + mRoomId + "] stopMusic:[id:" + id + "]");
        mTRTCCloud.getAudioEffectManager().stopPlayMusic(id);
        mMusicPanelState.currentMusicInfo.get().isPlaying.set(false);
    }

    public void deleteMusic(MusicPanelState.MusicInfo musicInfo) {
        boolean isPlaying = musicInfo.isPlaying.get();
        LiveKitLog.info(TAG + "[" + mRoomId + "] deleteMusic:[musicInfo:" + musicInfo
                + ",isPlaying:" + isPlaying + "]");
        if (isPlaying) {
            stopMusic(musicInfo.id);
        }
        mMusicPanelState.currentMusicInfo.set(new MusicPanelState.MusicInfo());
        mMusicPanelState.musicList.remove(musicInfo);
    }
}

File: Flutter/livekit/lib/component/music/service/music_service.dart
import 'package:tencent_live_uikit/common/language/gen/livekit_localizations.dart';
import 'package:tencent_live_uikit/common/widget/global.dart';
import 'package:tencent_trtc_cloud/trtc_cloud_def.dart';

import '../../../live_stream/manager/index.dart';
import '../store/music_state.dart';

class MusicService {
  late LiveController liveController;
  late MusicState musicState;

  MusicService({
    required this.liveController,
    required this.musicState,
  }) {
    _initData();
  }

  void operatePlayMusic(MusicInfo musicInfo) {
    final MusicInfo? currentMusicInfo = musicState.currentMusicInfo.value;
    if (currentMusicInfo?.isPlaying.value == true) {
      _stopMusic(currentMusicInfo!);
    }
    _startMusic(musicInfo);
  }

  void operateStopMusic(MusicInfo musicInfo) {
    if (musicState.currentMusicInfo.value == musicInfo) {
      musicState.currentMusicInfo.value = null;
    }
    _stopMusic(musicInfo);
  }

  void operateDeleteMusic(MusicInfo musicInfo) {
    if (musicState.currentMusicInfo.value == musicInfo) {
      musicState.currentMusicInfo.value = null;
    }
    if (musicInfo.isPlaying.value) {
      _stopMusic(musicInfo);
    }
    _deleteMusic(musicInfo);
  }
}

extension MusicServiceLogicExtension on MusicService {
  void _startMusic(MusicInfo musicInfo) {
    musicState.currentMusicInfo.value = musicInfo;
    musicInfo.isPlaying.value = true;
    final AudioMusicParam musicParam =
        AudioMusicParam(path: musicInfo.path, id: musicInfo.id, publish: true, loopCount: 100000000);
    liveController.liveService.startPlayMusic(musicParam);
  }

  void _stopMusic(MusicInfo musicInfo) {
    musicInfo.isPlaying.value = false;
    liveController.liveService.stopPlayMusic(musicInfo.id);
  }

  void _deleteMusic(MusicInfo musicInfo) {
    final List<MusicInfo> list = [];
    list.addAll(musicState.musicList.value);
    list.remove(musicInfo);
    musicState.musicList.value = list;
  }

  void _initData() {
    if (musicState.musicList.value.isEmpty) {
      List<MusicInfo> list = [];

      list.add(MusicInfo(
          id: 1,
          name: LiveKitLocalizations.of(Global.appContext())!.live_music_cheerful,
          path: "https://dldir1.qq.com/hudongzhibo/TUIKit/resource/music/PositiveHappyAdvertising.mp3"));
      list.add(MusicInfo(
          id: 2,
          name: LiveKitLocalizations.of(Global.appContext())!.live_music_melancholy,
          path: "https://dldir1.qq.com/hudongzhibo/TUIKit/resource/music/SadCinematicPiano.mp3"));
      list.add(MusicInfo(
          id: 3,
          name: LiveKitLocalizations.of(Global.appContext())!.live_music_wonder_world,
          path: "https://dldir1.qq.com/hudongzhibo/TUIKit/resource/music/WonderWorld.mp3"));
      musicState.musicList.value = list;
    }
  }
}
